driver-core/src/main/com/mongodb/internal/connection/OidcAuthenticator.java

```diff
@@ -0,0 +1,682 @@
+            Integer accessTokenExpiresInSeconds = idpResponse.getAccessTokenExpiresInSeconds();
+            if (accessTokenExpiresInSeconds != null) {
+                this.accessToken = idpResponse.getAccessToken();
+                long accessTokenExpiryReservedSeconds = TimeUnit.MINUTES.toSeconds(5);
+                this.accessTokenExpiry = Timeout.startNow(
+                        Math.max(0, accessTokenExpiresInSeconds - accessTokenExpiryReservedSeconds),
+                        TimeUnit.SECONDS);
+            } else {
+                this.accessToken = null;
+                this.accessTokenExpiry = null;
+            }
+            String refreshToken = idpResponse.getRefreshToken();
+            if (refreshToken != null) {
+                this.refreshToken = refreshToken;
+                this.idpInfo = idpInfo;
+            } else {
+                this.refreshToken = null;
+                this.idpInfo = null;
+            }
+        }
+
+        OidcCacheEntry() {
+            this(null, null, null, null);
+        }
+
+        private OidcCacheEntry(@Nullable final String accessToken, @Nullable final Timeout accessTokenExpiry,
+                @Nullable final String refreshToken, @Nullable final IdpInfo idpInfo) {
+            this.accessToken = accessToken;
+            this.accessTokenExpiry = accessTokenExpiry;
+            this.refreshToken = refreshToken;
+            this.idpInfo = idpInfo;
+        }
+
+        @Nullable
+        String getValidCachedAccessToken() {
+            if (accessToken == null || accessTokenExpiry == null || accessTokenExpiry.expired()) {
+                return null;
+            }
+            return accessToken;
+        }
+
+        @Nullable
+        String getRefreshToken() {
+            return refreshToken;
+        }
+
+        @Nullable
+        IdpInfo getIdpInfo() {
+            return idpInfo;
+        }
+
+        OidcCacheEntry clearAccessToken() {
+            return new OidcCacheEntry(
+                    null,
+                    null,
+                    this.refreshToken,
+                    this.idpInfo);
+        }
+
+        OidcCacheEntry clearRefreshToken() {
+            return new OidcCacheEntry(
+                    this.accessToken,
+                    this.accessTokenExpiry,
+                    null,
+                    null);
+        }
+    }
+
+    private final class OidcSaslClient extends SaslClientImpl {
+
+        private OidcSaslClient(final MongoCredentialWithCache mongoCredentialWithCache) {
+            super(mongoCredentialWithCache.getCredential());
+        }
+
+        @Override
+        public byte[] evaluateChallenge(final byte[] challenge) {
+            return evaluateChallengeInternal(challenge);
+        }
+
+        @Override
+        public boolean isComplete() {
+            return clientIsComplete();
+        }
+
+        public byte[] evaluateChallengeInternal(final byte[] challenge) {
```

Comment:
I am not sure why this method was extracted from `evaluateChallenge`, and I suggest to inline it. But if not, at the very least we should make it `private`.

```suggestion
        private byte[] evaluateChallengeInternal(final byte[] challenge) {
```